src/Data/Discrimination/Internal.hs

```diff
@@ -55,3 +64,21 @@ spanEither k xs0 = go [] xs0 where
 fromRight :: Either a b -> b
 fromRight (Right y) = y
 fromRight _ = error "unstable discriminator"
+
+-------------------------------------------------------------------------------
+-- * Integer and Natural
+-------------------------------------------------------------------------------
+
+integerCases :: Integer -> Either (GmpSize,[GmpLimb]) (Either Int (GmpSize,[GmpLimb]))
+integerCases (Jn# b) = Left          $ decomposeBigNat b
+integerCases (S#  i) = Right . Left  $ I# i
+integerCases (Jp# b) = Right . Right $ decomposeBigNat b
+
+naturalCases :: Natural -> Either GmpLimb (GmpSize,[GmpLimb])
+naturalCases (NatS# w#) = Left $ W# w#
+naturalCases (NatJ# b)  = Right $ decomposeBigNat b
+
+-- We need to reverse the limb array. Its stored least-significant word first
+-- but for comparasion to work right we need most-significant words first.
+decomposeBigNat :: BigNat -> (GmpSize, [GmpLimb])
+decomposeBigNat (BN# ba#) = let pa = PrimArray ba# :: PrimArray Word in (sizeofPrimArray pa, reverse $ primArrayToList pa)
```

Comment:
Would there be any benefit in avoiding the reverse in here? Building the list from the back seems pretty simple to do, I assume there's no optimisation for `reverse (build f)`?

Reply:
Good point. Changed